KAFKA-6859: Do not send LeaderEpochRequest for undefined leader epochs

stanislavkozlovski opened the pull request:
If a broker or topic has a message format < 0.11, he does not track leader epochs. LeaderEpochRequests for such will always return undefined, making the follower truncate to the highest watermark. Since there is no use to use the network for such cases, don't send a request.

### Notes
* I am not sure whether we should even check `brokerConfig.interBrokerProtocolVersion >= KAFKA_0_11_0_IV2`, since log message format is the only thing we care about and as far as I understand it is always defined
* I am not fond of the subclassing in the tests for mocking the method (and also making it protected) but this was the cleanest solution I could come up with. Other ideas are welcome

The ticket also mentions

> Another example is a bootstrapping follower that does not have any leader epochs recorded, 

but that is handled here:
https://github.com/apache/kafka/blob/57320981bb98086a0b9f836a29df248b1c0378c3/core/src/main/scala/kafka/server/AbstractFetcherThread.scala#L127
as far as I can tell

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

stanislavkozlovski committed:
KAFKA-6859: Do not send LeaderEpochRequest for undefined leader epochs

If a broker or topic has a message format < 0.11, he does not track leader epochs. LeaderEpochRequests for such will always return undefined, making the follower truncate to the highest watermark. Since there is no use to use the network for such cases, don't send a request.

stanislavkozlovski committed:
KAFKA-6859 - invert wrong comparison

stanislavkozlovski commented:
@apovzner, could you review this? (as you reported the ticket)

stanislavkozlovski commented:
(passed system tests) - https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1832/parameters/

apovzner commented:
The case where a bootstrapping follower sends OffsetsForLeaderEpoch with UNDEFINED_EPOCH can happen because ReplicaFetcherThread#buildLeaderEpochRequest() method will create OffsetsForLeaderEpoch for any partition whose epoch cache is initialized.
```
val (partitionsWithEpoch, partitionsWithoutEpoch) = partitionEpochOpts.partition { case (_, epochCacheOpt) => epochCacheOpt.nonEmpty }
```
However, latestEpoch() can return UNDEFINED_EPOCH if there are no entries in the cache.
```
val result = partitionsWithEpoch.map { case (tp, epochCacheOpt) => tp -> epochCacheOpt.get.latestEpoch() }
```
I am pretty sure we have seen that happen, but maybe double check.

stanislavkozlovski committed:
KAFKA-6859: Ensure bootstrapping follower does not send OffsetsForLeaderEpoch requests when he doesn't have epochs

LeaderEpochRequests sent with UNDEFINED_EPOCH will always return UNDEFINED_EPOCH, making the follower truncate to the highest watermark. Since bootstrapping follower brokers never have the epochs initialized in the cache, filter out those partitions and do not send a request for them

stanislavkozlovski commented:
Seems like it is possible:
https://github.com/apache/kafka/blob/8c47a3e52f3d6871a708b1d2c8b5a7e30a2a5b99/core/src/main/scala/kafka/server/epoch/LeaderEpochFileCache.scala#L79

Thanks for clarifying and pointing me in the right direction. 
Could you give it another look @apovzner?

stanislavkozlovski commented:
Retest this please

apovzner commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -343,16 +345,20 @@ class ReplicaFetcherThread(name: String,
-    val (partitionsWithEpoch, partitionsWithoutEpoch) = partitionEpochOpts.partition { case (_, epochCacheOpt) => epochCacheOpt.nonEmpty }
+    val (partitionsWithCache, partitionsWithoutCache) = partitionEpochOpts.partition { case (_, epochCacheOpt) => epochCacheOpt.nonEmpty }
+
+    debug(s"Build leaderEpoch request $partitionsWithCache")
```
Since leader epoch request does not include partitions with undefined epoch anymore, should this log output `$partitionsWithoutEpoch` instead (and move debug statement below next statement?

apovzner commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -378,6 +384,26 @@ class ReplicaFetcherThread(name: String,
+      return false
+
+    // if all topics' message format do not support epoch requests, skip it
+    partitions.map { case (tp, _) => topicSupportsEpochRequest(tp) }.reduce(_ || _)
```
If we are going to skip partitions on older message format, we could do that in `buildLeaderEpochRequest()` function, where we already skip partitions with no recorded leader epoch or partitions without epoch cache. I think you are right about the need to check message format, and possibly not requiring to check inter-broker protocol in that case, but maybe @junrao  could confirm.

stanislavkozlovski committed:
KAFKA-6859: Do not include partitions who do not support leader epoch in leader epoch request

* Also fixes tests

apovzner commented:
One of the test failures (JDK 10) seem to be related to this PR:
kafka.server.ReplicaManagerTest.testBecomeFollowerWhenLeaderIsUnchangedButMissedLeaderUpdate
It uses mocks for various fetcher thread calls. Since this PR changes `ReplicaFetcherThread.fetchEpochsFromLeader` to eventually call `ReplicaManager.getLog`, the test fails because of unexpected call to ReplicaManager.getLog:
```
[2018-08-20 11:13:50,602] ERROR [ReplicaFetcher replicaId=0, leaderId=1, fetcherId=0] Error due to (kafka.server.ReplicaManagerTest$$anon$4$$anon$2$$anon$1:76)
java.lang.AssertionError: 
  Unexpected method call LogManager.getLog$default$2():
    LogManager.truncateTo(Map(test-topic-0 -> 5), false): expected: 1, actual: 0
	at org.easymock.internal.MockInvocationHandler.invoke(MockInvocationHandler.java:44)
	at org.easymock.internal.ObjectMethodsFilter.invoke(ObjectMethodsFilter.java:102)
	at org.easymock.internal.ClassProxyFactory$MockMethodInterceptor.intercept(ClassProxyFactory.java:97)
	at kafka.log.LogManager$$EnhancerByCGLIB$$a9c8c911.getLog$default$2(<generated>)
	at kafka.server.ReplicaManager.getLog(ReplicaManager.scala:278)
	at kafka.server.ReplicaFetcherThread.topicSupportsEpochRequest(ReplicaFetcherThread.scala:402)
	at kafka.server.ReplicaFetcherThread.$anonfun$shouldSendLeaderEpochRequest$1(ReplicaFetcherThread.scala:397)
	at kafka.server.ReplicaFetcherThread.$anonfun$shouldSendLeaderEpochRequest$1$adapted(ReplicaFetcherThread.scala:397)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
	at scala.collection.immutable.Map$Map1.foreach(Map.scala:125)
	at scala.collection.TraversableLike.map(TraversableLike.scala:234)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:227)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at kafka.server.ReplicaFetcherThread.shouldSendLeaderEpochRequest(ReplicaFetcherThread.scala:397)
	at kafka.server.ReplicaFetcherThread.fetchEpochsFromLeader(ReplicaFetcherThread.scala:362)
	at kafka.server.AbstractFetcherThread.maybeTruncate(AbstractFetcherThread.scala:133)
	at kafka.server.AbstractFetcherThread.doWork(AbstractFetcherThread.scala:105)
	at kafka.server.ReplicaManagerTest$$anon$4$$anon$2$$anon$1.doWork(ReplicaManagerTest.scala:705)
	at kafka.utils.ShutdownableThread.run(ShutdownableThread.scala:82)
```

hachikuji commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -22,7 +22,7 @@ import java.util
 import kafka.api._
 import kafka.cluster.BrokerEndPoint
-import kafka.log.LogConfig
+import kafka.log.{Log, LogConfig}
```
nit: unused?

hachikuji commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -95,7 +95,9 @@ class ReplicaFetcherThread(name: String,
   private val fetchSize = brokerConfig.replicaFetchMaxBytes
-  private val shouldSendLeaderEpochRequest: Boolean = brokerConfig.interBrokerProtocolVersion >= KAFKA_0_11_0_IV2
+  private val brokerSupportsLeaderEpochRequest: Boolean = brokerConfig.interBrokerProtocolVersion >= KAFKA_0_11_0_IV2 &&
+    brokerConfig.logMessageFormatVersion >= KAFKA_0_11_0_IV2
```
It is possible for the default message format version to be lower than what is configured for the topic, so I'm not sure this addition makes sense.

apovzner commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -343,16 +345,21 @@ class ReplicaFetcherThread(name: String,
+    val (partitionsWithCache, partitionsWithoutCache) = partitionEpochOpts.partition { case (_, epochCacheOpt) => epochCacheOpt.nonEmpty }
+    val (partitionsWithEpoch, partitionsWithoutEpoch) = partitionsWithCache
+      .map { case (tp, epochCacheOpt) => tp -> epochCacheOpt.get.latestEpoch() }
+      .partition { case (_, latestEpoch ) => latestEpoch != UNDEFINED_EPOCH }
```
I spent a bit more time looking at this PR and the code. If we add partitions that do not have any leader epoch recorded in the leader epoch cache to the `partitionsWithError` set returned from this method, those partitions will be delayed. I think they will end up being delayed indefinitely. So, my previous comment that we can also try to filter partitions on older message format in this method is incorrect, I apologize. 
I think we need to keep this method unchanged (pre-this PR). I understand that partitions without epoch cache are counted as "partitions with error", because we wait for the cache to be loaded. Other filtering needs to be done later, and making sure that partitions proceed to the truncation stage. 
It's quite worrying that we don't have unit/integration test that would catch that this change may cause some partitions to not move to truncate and then fetching state.

apovzner commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -92,6 +135,100 @@ class ReplicaFetcherThreadTest {
+      }
+    }
+    // topic 1 supports epoch, t2 doesn't
+    thread.addPartitions(Map(t1p0 -> 0, t1p1 -> 0, t2p1 -> 0))
```
related to my other comment, this test needs to verify that partitions on older message format (that do not send request for leader epoch) proceed to truncation state and to fetch state. Since we batch several partitions into the same request for leader epoch and fetch request, verifying that we sent one request for leader epoch or one fetch request during one `doWork` loop is not enough. We also need to verify that every partition moved to fetching state.

stanislavkozlovski replied:
I reverted `buildLeaderEpochRequest` to its original implementation, but before that I verified that the new test code catches that it hasn't moved to a fetch state

hachikuji commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -343,16 +345,21 @@ class ReplicaFetcherThread(name: String,
-    debug(s"Build leaderEpoch request $partitionsWithEpoch")
-    val result = partitionsWithEpoch.map { case (tp, epochCacheOpt) => tp -> epochCacheOpt.get.latestEpoch() }
-    ResultWithPartitions(result, partitionsWithoutEpoch.keys.toSet)
+    ResultWithPartitions(partitionsWithEpoch, partitionsWithError)
```
I think the desired behavior is for partitions which do not support the latest message format to fall back to truncation using the high watermark. However, `partitionsWithError` ultimately get passed to `handlePartitionsWithErrors`, which then calls `delayPartitions`. That seems not quite what we want?

hachikuji commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -378,6 +385,26 @@ class ReplicaFetcherThread(name: String,
+  // protected for test mocking
+  protected def topicSupportsEpochRequest(tp: TopicPartition): Boolean = {
+    replicaMgr.getLog(tp) match {
+      case Some(log) => log.config.messageFormatVersion >= KAFKA_0_11_0_IV2
```
There is a little bit of trickiness here in that the leader may have its message format version updated before the follower. However, the follower always accepts the message format of the leader. I'm wondering if we should skip the config check and send the OffsetsForLeaderEpoch request iff we have an epoch to query?

junrao replied:
Agreed that this approach sounds right.

hachikuji reviewed (commented):
Thanks for the PR. Left a few comments.

stanislavkozlovski committed:
KAFKA-6859: Address code review comments

* Do not look at log message format version when deciding whether it supports epoch requests
* Do not send leader epoch request for partitions when the answer is known (undefined epoch)
* Validate expected states in ReplicaFetcherThreadTest

stanislavkozlovski committed:
KAFKA-6859: Change LeaderEpochCache#latestEpoch to return Option[Int]

stanislavkozlovski commented:
We discussed offline with @hachikuji and decided that it would be better if `LeaderEpochCache#latestEpoch()` would return an `Option[Int]`, rather than a sentinel value of -1 for an undefined epoch. 
I didn't completely remove the `UNDEFINED_EPOCH` notion throughout the code but nevertheless changed the method signature

hachikuji commented on core/src/test/scala/unit/kafka/server/epoch/LeaderEpochFileCacheTest.scala:
```diff
@@ -48,7 +48,7 @@ class LeaderEpochFileCacheTest {
 
     //Then
-    assertEquals(2, cache.latestEpoch())
+    assertEquals(2, cache.latestEpoch.get)
```
nit: we could change these assertions to something like this:
```scala
assertEquals(Some(2), cache.latestEpoch)
```

hachikuji commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -358,7 +358,7 @@ class LogSegment private[log] (val log: FileRecords,
         if (batch.magic >= RecordBatch.MAGIC_VALUE_V2) {
           leaderEpochCache.foreach { cache =>
-            if (batch.partitionLeaderEpoch > cache.latestEpoch()) // this is to avoid unnecessary warning in cache.assign()
+            if (batch.partitionLeaderEpoch > cache.latestEpoch.getOrElse(-1)) // this is to avoid unnecessary warning in cache.assign()
```
Another way to write this:
```scala
cache.latestEpoch.forall(batch.partitionLeaderEpoch > _)
```

Also, can we drop the parethesis from the definition of `latestEpoch` in `LeaderEpochCache`?

stanislavkozlovski replied:
Does that actually result in clearer code? I find that harder to read than the above

Dropped the parenthesis 👍

hachikuji replied:
It's a bit subjective for sure, so feel free to ignore the suggestion. I just thought the use of -1 was a little odd. A more explicit alternative might be something like this:
```
if (latestEpoch.isEmpty || batch.partitionLeaderEpoch > latestEpoch.get)
```
I'll leave it up to you.

hachikuji commented on core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala:
```diff
@@ -156,10 +156,10 @@ class ReplicaAlterLogDirsThread(name: String,
     partitions.map { case (tp, epoch) =>
       try {
-        val (leaderEpoch, leaderOffset) = replicaMgr.getReplicaOrException(tp).epochs.get.endOffsetFor(epoch)
+        val (leaderEpoch, leaderOffset) = replicaMgr.getReplicaOrException(tp).epochs.get.endOffsetFor(epoch.getOrElse(UNDEFINED_EPOCH))
```
It would be nice if we could limit use of `UNDEFINED_EPOCH` to the serde logic. I think passing around integer sentinels is unsafe because the type checker can't help us ensure that we have proper checking in place. A couple options come to mind:

1. We can let `endOffsetFor` accept `Option[Int]` as the epoch to query. The epoch if defined must be non-negative.
2. Keep the current `endOffsetFor` API, but require the epoch to be non-negative. We can match on the epoch here and only invoke `endOffsetFor` if we have an epoch. Otherwise we can return `(UNDEFINED_EPOCH, UNDEFINED_EPOCH_OFFSET` directly.

stanislavkozlovski replied:
Not sure which one is better, but I guess 2. might be better - the less `options` the simpler the code, right?
Do you think it is worth checking and raising an exception in `endOffsetFor` if a negative epoch is sent? I went without it

hachikuji commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -338,27 +339,31 @@ class ReplicaFetcherThread(name: String,
+    if (shouldSendLeaderEpochRequest(partitions)) {
+      // skip request for partitions without epoch, as their topic log message format doesn't support epochs
+      val (partitionsWithEpoch, partitionsWithoutEpoch) = partitions.partition { case (_, epoch) => epoch.nonEmpty }
+      val partitionsAsJava = partitionsWithEpoch.map { case (tp, epoch) => tp -> epoch.get.asInstanceOf[Integer] }.toMap.asJava
```
I think we can simplify this a little bit. If `partitionsWithEpoch` is empty, then we can just skip the send? Then we don't need the changes to `shouldSendLeaderEpochRequest`.

hachikuji reviewed (commented):
Thanks for the updates. Left a few more comments.

stanislavkozlovski committed:
KAFKA-6859: Code review comments

stanislavkozlovski committed:
KAFKA-6859: Switch #fetchEpochsFromLeader to use a guard clause

stanislavkozlovski commented:
@hachikuji I think I addressed your comments.
Did a bit of experimentation with the code structure in this commit - 1f595ccb20c700d48dcb6713274fa7100528db8f
I feel it looks a bit better, wdyt? Are "guard clauses" common in scala?

stanislavkozlovski committed:
KAFKA-6859: Fix tests

hachikuji commented on core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala:
```diff
@@ -156,10 +156,17 @@ class ReplicaAlterLogDirsThread(name: String,
       try {
-        val (leaderEpoch, leaderOffset) = replicaMgr.getReplicaOrException(tp).epochs.get.endOffsetFor(epoch)
+        val (leaderEpoch, leaderOffset) =
+          if (epoch.nonEmpty)
```
Could be a `match`?

hachikuji commented on core/src/main/scala/kafka/server/epoch/LeaderEpochFileCache.scala:
```diff
@@ -92,17 +92,13 @@ class LeaderEpochFileCache(topicPartition: TopicPartition, leo: () => LogOffsetM
-          // a follower is on the older message format where leader epochs are not recorded
-          (UNDEFINED_EPOCH, UNDEFINED_EPOCH_OFFSET)
-        } else if (requestedEpoch == latestEpoch) {
+        if (requestedEpoch == latestEpoch.getOrElse(UNDEFINED_EPOCH)) {
```
`latestEpoch.contains(requestedEpoch)`?

hachikuji commented on core/src/main/scala/kafka/server/epoch/LeaderEpochFileCache.scala:
```diff
@@ -92,17 +92,13 @@ class LeaderEpochFileCache(topicPartition: TopicPartition, leo: () => LogOffsetM
     * so that the follower falls back to High Water Mark.
     *
-    * @param requestedEpoch requested leader epoch
+    * @param requestedEpoch requested leader epoch. Must be non-negative
```
One use which seems unprotected is `Partition.lastOffsetForLeaderEpoch`. Seems like it will take a bit more work to purge the use of UNDEFINED_EPOCH.. Unless you're up for following through the full usage, maybe we had better leave the check for UNDEFINED_EPOCH in place.

stanislavkozlovski replied:
We did 70% of the work, might as well do the last 30%. As far as I can tell, they're only 4 calls to that method (3 of which from tests), all go through `ReplicaManager#lastOffsetForLeaderEpoch`

hachikuji replied:
Can we raise an exception if the epoch is non-negative? Also, I think the docstring above needs to be updated.

hachikuji reviewed (commented):
Thanks for the updates. The logic for skipping the request looks good. I think my previous suggestion about `LeaderEpochCache.endOffsetFor` may have been a little overeager, but let me know what you think.

stanislavkozlovski committed:
KAFKA-6859: Change Partition#lastOffsetForLeaderEpoch to accept Option[Int]

This is due to limiting the usage of the UNDEFINED_EPOCH integer sentinel (-1)

hachikuji commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -720,12 +720,16 @@ class Partition(val topic: String,
         case Some(leaderReplica) =>
-          val (epoch, offset) = leaderReplica.epochs.get.endOffsetFor(leaderEpoch)
-          new EpochEndOffset(NONE, epoch, offset)
+          if (leaderEpoch.isEmpty)
```
nit: use a `match`?

hachikuji commented on core/src/main/scala/kafka/server/epoch/LeaderEpochFileCache.scala:
```diff
@@ -74,9 +74,9 @@ class LeaderEpochFileCache(topicPartition: TopicPartition, leo: () => LogOffsetM
     * @return
     */
-  override def latestEpoch(): Int = {
+  override def latestEpoch: Option[Int] = {
```
nit: fill in the `@return` doc above?

hachikuji reviewed (commented):
Thanks, just a couple more comments.

stanislavkozlovski committed:
KAFKA-6859: Change EpochEndOffset to hold Optional of leader epoch/endoffset

stanislavkozlovski commented:
@hachikuji I also made `EpochEndOffset` hold optionals. I added a check inside the class to convert Some(-1) to None, since -1 can still get passed in the `OffsetsForLeaderEpochResponse`. 
Not sure if that can be changed to accept an optional, since the response is going to be an integer at the end of the day

Quite a lot of changes and I'm frankly not sure if I haven't missed anything. Tests pass locally but I am still not a 100% confident in the changes

stanislavkozlovski committed:
Revert LeaderEpoch Option changes

stanislavkozlovski commented:
@hachikuji Come to think of it, these changes are too much for this ticket. 
It'd be best if we continue them in a separate PR. I've reverted the use of Option[] from this PR and only left out the original changes.
When this gets merged, I'll re-open another PR substituting undefined_epoch with Option[] and we can see how we feel about that. Is that okay?

hachikuji commented:
retest this please

hachikuji commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -89,9 +100,148 @@ class ReplicaFetcherThreadTest {
+    *   and there is no use to send the request, as we know the broker will answer with that epoch
+    */
+  @Test
+  def shouldNotIssueLeaderEpochRequestIfIfBrokerProtocolAbove11AllTopicsDontSupportEpochs(): Unit = {
```
The name seems a little stale. Perhaps `shouldNotSendEpochRequestIfLastEpochUndefinedForAllPartitions` or something like that?  The other test case probably needs a similar update.

hachikuji commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -89,9 +100,148 @@ class ReplicaFetcherThreadTest {
+    * If a partition doesn't have an epoch in the cache, this means it either
+    *   does not support epochs (log message format below 0.11) or it is a bootstrapping follower.
+    *
+    * In both cases, he has an undefined epoch
```
nit: "he" -> "the partition"?

Personally, I've always thought of partitions as feminine. 😉

stanislavkozlovski replied:
Haha, love it. Thanks for catching that, I try to avoid gender-specific pronouns in code comments - I think it's a language-barrier issue

hachikuji reviewed (commented):
Thanks, just a couple nits on the test cases. I agree that the optional change ended up being a bit too heavy for this fix, but thanks for trying it out.

stanislavkozlovski committed:
Code review comments

hachikuji reviewed (approved):
LGTM. Thanks for the patch.

hachikuji committed:
Calls to leaderEpoch() needed updating

hachikuji commented:
Pushed a fix and rebased. Tests pass locally, so I will go ahead and merge.

stanislavkozlovski commented:
Thanks about that
